xdsl/dialects/arith.py

```diff
@@ -148,21 +149,55 @@ class ConstantOp(IRDLOperation):
     @overload
     def __init__(
         self,
-        value: IntegerAttr | FloatAttr[AnyFloat] | DenseIntOrFPElementsAttr,
-        value_type: None = None,
+        value: IntegerAttr,
+        value_type: IntegerType | None = None,
     ) -> None: ...
 
     @overload
-    def __init__(self, value: Attribute, value_type: Attribute) -> None: ...
+    def __init__(
+        self,
+        value: FloatAttr[AnyFloat],
+        value_type: AnyFloat | None = None,
+    ) -> None: ...
 
+    @overload
     def __init__(
         self,
-        value: IntegerAttr | FloatAttr[AnyFloat] | Attribute,
+        value: DenseIntOrFPElementsAttr,
+        value_type: RankedStructure[AnyFloat | IntegerType | IndexType] | None = None,
+    ) -> None: ...
+
+    @overload
+    def __init__(
+        self,
+        value: Attribute,
+        value_type: Attribute,
+    ) -> None: ...
+
+    def __init__(
+        self,
+        value: IntegerAttr | FloatAttr[AnyFloat] | DenseIntOrFPElementsAttr | Attribute,
         value_type: Attribute | None = None,
     ):
         if value_type is None:
-            value = cast(IntegerAttr | FloatAttr[AnyFloat], value)
-            value_type = value.type
+            if isinstance(value, IntegerAttr | FloatAttr):
+                value = cast(IntegerAttr | FloatAttr[AnyFloat], value)
+                value_type = value.type
+            elif isinstance(value, DenseIntOrFPElementsAttr):
+                value_type = value.type
+
+        if isinstance(value, IntegerAttr):
+            value_type = cast(IntegerType, value_type)
+            value = IntegerAttr(value.value, value_type)
```

Comment:
I am against doing a silent coercion here. Personally I think we should either throw an error, or not allow a type in the constructor.

Reply:
I am in favour of not allowing the type, as this is what is the most clear IMHO. I didn't do it as I wasn't sure of the initial rationale for this API and this provides the minimum change against it.

As for throwing an error, I don't think we are doing it anywhere in any constructor.

Reply:
I believe the original API allowed passing in python ints? I would be in favour of a single constructor that takes a single attribute and infers the type from that attribute.

Reply:
The original API? I'm confused, the diff is clear, no? I can't see what you're saying. We have the `from_int_and_width` for that.

Reply:
Sorry, I meant original as in "when arith.constant was first added to xDSL" original.

Reply:
Well, I can't answer this, and at the end of the day that was changed to the current one that uses a typed attribute and a type (for which I mentioned I don't know the rationale behind it, AFAIU when I checked a while back, it was another deprecation of similar helpers to `from_int_and_width`, e.g. `from_attr`).

Reply:
I think at this point we can just remove value_type from the argument

Reply:
There seem to be uses within the stencil part of xDSL that use the general `Attribute` overload and rely on passing a `value_type`; for the rest of the types the API is simpler without requiring a `value_type` provided, as discussed previously here.